### PR TITLE
Raise when packing is combined with context parallelism

### PR DESCRIPTION
**Pull request by qgallouedec**
`docs/source/distributing_training.md` currently recommends packing as a best practice for context parallelism:

> **Use packing with padding** - The default BFD (Best Fit Decreasing) strategy works perfectly:
> - Preserves sequence boundaries and maintains training quality

Sequence boundaries are **not** preserved under context parallelism. Packed documents are kept apart by a block-diagonal attention mask, and context parallelism can only express full causal attention, accelerate's CP hook replaces every layer's mask with `is_causal=True`. So each packed document attends to every document before it in the same sequence, silently.

### Repro

Weight-independent probe: perturb one token in the first packed document and count how many output positions change. Only the first document should be affected. The model uses plain full attention, so packing is the only thing under test.

```
torchrun --nproc_per_node=2 repro_cp_packing.py
```

```
no CP: 50.0% of positions change  (50% means the two documents are correctly separated)
CP=2 : 100.0%  (100% means the second document sees the first)
```

<details>
<summary><code>repro_cp_packing.py</code></summary>

```python
import argparse
import os

import torch
import torch.distributed as dist
from accelerate.big_modeling import _attach_context_parallel_hooks
from torch.distributed.device_mesh import init_device_mesh
from torch.distributed.tensor.experimental import context_parallel
from transformers import AutoModelForCausalLM, Qwen3Config

DOC = 256  # two packed documents of this length


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--dump", help="save the per-position change mask for plotting")
    args = parser.parse_args()

    rank = int(os.environ["RANK"])
    torch.cuda.set_device(rank)
    dist.init_process_group("nccl")
    device = torch.device("cuda", rank)
    mesh = init_device_mesh("cuda", (dist.get_world_size(),), mesh_dim_names=("cp",))

    config = Qwen3Config(vocab_size=1024, hidden_size=256, intermediate_size=512, num_hidden_layers=2,
                         num_attention_heads=4, num_key_value_heads=2, head_dim=64,
                         attn_implementation="sdpa")
    torch.manual_seed(0)
    model = AutoModelForCausalLM.from_config(config).to(device, torch.bfloat16).eval()

    ids_a = torch.randint(0, 1024, (1, 2 * DOC), device=device)
    ids_b = ids_a.clone()
    ids_b[0, 0] = (ids_a[0, 0] + 1) % 1024  # perturb the first token of the first document
    # restarting position_ids is what marks the document boundary
    positions = torch.cat([torch.arange(DOC), torch.arange(DOC)]).unsqueeze(0).to(device)

    def changed(cp):
        def logits(ids):
            kwargs = dict(input_ids=ids, position_ids=positions, use_cache=False)
            if not cp:
                with torch.no_grad():
                    return model(**kwargs).logits.float()
            with torch.no_grad(), context_parallel(mesh, buffers=[ids, positions], buffer_seq_dims=[1, 1]):
                return model(**kwargs).logits.float()

        delta = (logits(ids_a) - logits(ids_b)).abs().amax(-1)[0] > 1e-3
        if not cp:
            return delta
        gathered = [torch.zeros_like(delta) for _ in range(dist.get_world_size())]
        dist.all_gather(gathered, delta)
        return torch.cat(gathered)

    reference = changed(cp=False)
    _attach_context_parallel_hooks(model)
    with_cp = changed(cp=True)

    if rank == 0:
        print(f"no CP: {reference.float().mean():.1%} of positions change  "
              f"(50% means the two documents are correctly separated)")
        print(f"CP={dist.get_world_size()} : {with_cp.float().mean():.1%}  "
              f"(100% means the second document sees the first)")
        if args.dump:
            torch.save({"reference": reference.cpu(), "with_cp": with_cp.cpu(), "doc": DOC}, args.dump)
    dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

</details>

### The fix

Raise in `SFTTrainer.__init__` when `args.packing` and `parallelism_config.cp_enabled`, and correct the doc section. This matches how the other trainers handle context parallelism: `GRPOConfig`, `RLOOConfig` and `DistillationConfig` already raise for `cp_size > 1`.

Note that padding-free / flash-attention packing is unaffected — this only concerns the context-parallel path, which requires SDPA.

### Context

The root cause is in accelerate (its CP hook drops the mask without checking it); that side is addressed in https://github.com/huggingface/accelerate/pull/4177, which fixes the model-level (sliding-window) case. Packing is data-level, so it can only be caught here.

---

> [!NOTE]
> **Medium Risk**
> Changes only misconfigured CP+packing setups from silent wrong training to a hard error; valid configs are unchanged, but anyone relying on the old doc guidance must update their SFT config.
> 
> **Overview**
> **Prevents silent cross-document attention** when Ring Attention / context parallelism (`cp_size > 1`) is used with SFT packing.
> 
> `SFTTrainer` now raises a clear `ValueError` at init if Accelerate reports `cp_enabled` and either `packing` or `eval_packing` is on. The check uses `self.accelerator.parallelism_config` (not training args) so it still applies when CP is set only in an Accelerate YAML, and is gated on Accelerate ≥ 1.10.1.
> 
> The **distributing training** docs are updated to match: the Ring Attention example drops `packing=True`, and best practices now say **not** to use packing with CP because CP forces full causal attention and cannot preserve packed sequences’ block-diagonal masks.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8cc16d9230831a6c324d94b90c3f513397c9d87b. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1362,6 +1362,19 @@ def __init__(
             preprocess_logits_for_metrics=preprocess_logits_for_metrics,
         )
 
+        # Context parallelism can only express full causal attention: the per-layer attention mask is
+        # dropped and replaced by `is_causal=True`. Packed sequences rely on a block-diagonal mask to
+        # keep documents from attending to each other, so under context parallelism a packed batch
+        # would silently train with documents attending across their boundaries.
+        parallelism_config = self.accelerator.parallelism_config
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py]: Stale: since d7c60d7b the read is gated by an explicit version check (not _is_package_version_below("accelerate", "1.10.1")), so accelerate 1.4-1.10 never touches parallelism_config.

File: trl/trainer/sft_trainer.py [hunk 2]
```diff
@@ -1362,6 +1362,19 @@ def __init__(
...
+        if parallelism_config is not None and parallelism_config.cp_enabled and args.packing:
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply trl/trainer/sft_trainer.py]: Correct, covered in cb37381b: the condition is now (args.packing or args.eval_packing) and the error message names both.

File: docs/source/distributing_training.md [hunk 3]
```diff
@@ -204,9 +204,10 @@ accelerate launch --config_file context_parallel_2gpu.yaml train.py
    - For `cp_size=4`: use `pad_to_multiple_of=8` (since `cp_size * 2 = 8`)
    - The data collator automatically pads sequences to the required multiple, ensuring compatibility with CP
 
-2. **Use packing with padding** - The default BFD (Best Fit Decreasing) strategy works perfectly:
-   - Preserves sequence boundaries and maintains training quality
-   - Works seamlessly with both `padding_free=True` and standard padding modes
+2. **Do not use packing** - Context parallelism can only express full causal attention: the per-layer attention
+   mask is dropped and replaced by `is_causal=True`. Packed sequences rely on a block-diagonal mask to keep the
+   packed documents separate, so under context parallelism they would attend across document boundaries. TRL raises
+   an error for this combination.
```
[the remark this answers is not in the document]
Comment (qgallouedec) [reply docs/source/distributing_training.md:209]: Correct, fixed in cb37381b: packing=True is gone from the example config.

**Comment by qgallouedec** [comment]
merging without review, feel free to review later if you think I missed something
